Source/ARTClientOptions.h

```diff
@@ -134,6 +135,11 @@ NS_ASSUME_NONNULL_BEGIN
  */
 @property (nullable, weak, nonatomic) id<ARTPushRegistererDelegate, NSObject> pushRegistererDelegate;
 
+/**
+ Custom local device storage. In the case nothing is provided then a default implementation using `ARTDefaultLocalDeviceStorage` is used.
+ */
+@property (nullable, nonatomic) id<ARTDeviceStorage> storage;
```

Comment:
`setObject:` and `objectForKey:` are just fallbacks to `NSUserDefaults` `setObject:` and `objectForKey:`. Whereas `secret` uses either now public keychain methods (during migration) or also fallbacks to `NSUserDefaults`. All methods are synchronous. The security implications are simple: we store `secret` in a plain text. We must warn users that it's probably fine for prototypes but can be harmful for production code, and encourage them to use something more secure. Implementing reliable and secure mechanism for storing data is a whole new universe and way beyond of the realm of this PR.

Reply:
I think that the answers to these questions need to be documented alongside the protocol.

>  We must warn users that it's probably fine for prototypes but can be harmful for production code, and encourage them to use something more secure. 

If our default "can be harmful for production code" then we need to explain why it "can be harmful" — i.e. what are the potential consequences of a leak of a token (see my final bullet point in my question).

Reply:
I guess potentially using users generated device id/secret anyone can publish/receive messages on a behalf of said user. WDYT @ben-xD

Reply:
I think it's harmful in some cases: e.g. jailbroken phones? It requires some investigation and then documentation.

Reply:
>I think it's harmful in some cases: e.g. jailbroken phones?

Yes, I think it's physical access to the phone, jailbrake and Security Services stuff like Pegasus and such. But in those cases leaked device secret is probably not the first thing user should be concerned about.

Reply:
I just read https://stackoverflow.com/questions/8717419/ios-is-it-secure-to-store-sensible-user-information-in-nsuserdefaults-standard, and it is also debatable. Some say you need physical access to the device, and another says [you're okay to store authentication tokens in User defaults](https://stackoverflow.com/a/8717491/7365866).

I think it's worth adding something to the documentation related to push notifications: e.g.
> DeviceId, DeviceSecret, Device Identity Token and APNs Device Token are saved in UserDefaults. This means in environments where UserDefaults is compromised (e.g. jailbroken device), this information can be leaked. To enhance security, implement `ARTDeviceStorage` which uses the keychain. 

However, I am not comfortable to ask users to use `ARTKeychainLocalDeviceStorage`, since this might be the cause of the errors we see in the Realtime serverss.

Reply:
@lawrence-forooghian WDYT about the comment in documentation to the `ARTDeviceStorage` provided by @ben-xD 
Is it enough to push this PR forward?

>Ably Cocoa saves the APNs Device Token and Ably specific values (Device ID, Device Secret and Device Identity Token) in User Defaults. This means in environments where User Defaults is compromised (e.g. jailbroken device), this information can be leaked. To enhance security, create your own implementation of ARTDeviceStorage which uses the keychain.

Reply:
Makes sense, I think. I'm wondering about the "To enhance security, create your own implementation of ARTDeviceStorage which uses the keychain." bit — if we start telling users to use the keychain, mightn't they just end up with the same issue that we are trying to fix here (i.e. unable to fetch from keychain, leading to a new device registration)?

Reply:
I bet they will end up there for sure. We should probably encourage them to use `UserDefaults` default implementation then and point out, that it is safe if the device is not jailbroken. WDYT?

Reply:
Don't give up so easily Marat 😅. This part can be improved to:
>  Ably Cocoa saves the APNs Device Token and Ably specific values (Device ID, Device Secret and Device Identity Token) in **[User Defaults](https://developer.apple.com/documentation/foundation/userdefaults)**. This means in environments where User Defaults is compromised (e.g. jailbroken device), this information can be leaked. To enhance security, create your own implementation of `ARTDeviceStorage` which uses a more secure mechanism.

Reply:
I think it's better, because it will reduce the amount of users which went with keychain implementation. Probably 😅